pandas/tseries/offsets.py

```diff
@@ -603,61 +634,66 @@ def next_bday(self):
         else:
             return BusinessDay(n=nb_offset)
 
-    @cache_readonly
-    def _get_daytime_flag(self):
-        if self.start == self.end:
-            raise ValueError('start and end must not be the same')
-        elif self.start < self.end:
-            return True
-        else:
-            return False
+    def _get_daytime_flag(self, start, end):
+        return start < end
 
-    def _next_opening_time(self, other):
+    def _next_opening_time(self, other, dir=1):
         """
         If n is positive, return tomorrow's business day opening time.
         Otherwise yesterday's business day's opening time.
 
         Opening time always locates on BusinessDay.
         Otherwise, closing time may not if business hour extends over midnight.
         """
+        earliest_start = self.start[0]
+        latest_start = self.start[-1]
         if not self.next_bday.onOffset(other):
-            other = other + self.next_bday
+            other = other + dir * self.next_bday
+            if self.n * dir >= 0:
+                return datetime(other.year, other.month, other.day,
+                                earliest_start.hour, earliest_start.minute)
+            else:
+                return datetime(other.year, other.month, other.day,
+                                latest_start.hour, latest_start.minute)
+        else:
+            if self.n * dir >= 0 and latest_start < other.time():
+                other = other + dir * self.next_bday
+                return datetime(other.year, other.month, other.day,
+                                earliest_start.hour, earliest_start.minute)
+            elif self.n * dir < 0 and other.time() < earliest_start:
+                other = other + dir * self.next_bday
+                return datetime(other.year, other.month, other.day,
+                                latest_start.hour, latest_start.minute)
+        if self.n * dir >= 0:
+            for st in self.start:
+                if other.time() <= st:
+                    return datetime(other.year, other.month, other.day,
+                                    st.hour, st.minute)
```

Comment:
Comment your code to explain what all of this logic is for.

Reply:
Comments added